web-client/public/js/graph.js

```diff
@@ -93,74 +93,98 @@ var drawGraph = function (nodes, links, positiveWeights, negativeWeights, sheetT
         return Math.round(totalScale(Math.abs(edge.value)));
     };
 
-    var force = d3.layout.force()
-      .size([width, height])
-      .on("tick", tick)
-      .linkDistance($("#linkDistInput").val())
-      .charge($("#chargeInput").val())
-      .chargeDistance($("#chargeDistInput").val())
-      .gravity($("#gravityInput").val());
+    var simulation = d3.forceSimulation()
+      .force("link", d3.forceLink())
+      .force("charge", d3.forceManyBody())
+      .force("center", d3.forceCenter(width / 2, height / 2));
+
+    var drag = d3.drag()
+      .on("start", dragstart)
+      .on("drag", dragged)
+      .on("end", dragended);
+
+    var dragended = function () {
+        d3.event.stopPropagation();
+    };
+
+    var zoomDragPrevX = 0;
+    var zoomDragPrevY = 0;
+    var zoomDragStarted = function () {
+        zoomDragPrevX = d3.event.x;
+        zoomDragPrevY = d3.event.y;
+        $container.removeClass(CURSOR_CLASSES).addClass("cursorGrabbing");
+        if (!adaptive) {
+            $container.removeClass(CURSOR_CLASSES);
+        }
+    };
+
+    var zoomDragged = function () {
+        if (adaptive) {
+            var scale = 1;
+            if (zoomContainer.attr("transform")) {
+                var string = zoomContainer.attr("transform");
+                scale = 1 / (+(string.substring(string.indexOf("scale(") + 6, string.lastIndexOf(")"))));
```

Comment:
This will be a tad more compact (and slightly more readable) with a regex capturing a group:
```javascript
scale = 1 / +(string.match(/scale\(([^\)]+)\)/)[1]);
```

(uhhhh test that first)

Reply:
Yes, more readable and works fine!